File: mobile-imu-core/harness-ios/Sources/SampleRecorder.swift
import CoreMotion
import Foundation

/// Throwaway capture-only recorder for U0 (see
/// docs/plans/2026-08-21-001-feat-phone-imu-pendulum-app-plan.md).
///
/// Writes one JSON array of samples per trial, each shaped as
/// {t, role, sensor, v, phone_ts_ms} to match imu_calibration_tuner.py's
/// replay_trial() input contract. `role` is fixed to "proximal" — that is
/// the role replay_trial()'s solo-fallback path (is_solo = not has_distal
/// and role == ROLE_PROXIMAL) expects for a single-phone capture; a "distal"
/// role with no proximal counterpart would NOT trigger that fallback.
///
/// No calibration, no scoring, no participant model. This is not reused by
/// U3-U13 — see U0's plan entry for why.
final class SampleRecorder {
    private struct Sample {
        let tSeconds: Double
        let sensor: String
        let v: (Double, Double, Double)
        let phoneTsMs: Int64
    }

    private let motionManager = CMMotionManager()
    private var samples: [Sample] = []
    private var startUptime: TimeInterval?

    var sampleCount: Int { samples.count }

    /// ~100Hz, matching KTD5 — CMMotionManager's raw streams, not the newer
    /// CMBatchedSensorManager 800Hz path, since pendulum motion is low-frequency.
    private let updateInterval: TimeInterval = 1.0 / 100.0

    func start() {
        samples.removeAll()
        startUptime = nil
        guard motionManager.isAccelerometerAvailable,
              motionManager.isGyroAvailable,
              motionManager.isMagnetometerAvailable else {
            return
        }

        motionManager.accelerometerUpdateInterval = updateInterval
        motionManager.gyroUpdateInterval = updateInterval
        motionManager.magnetometerUpdateInterval = updateInterval

        // Raw accelerometerData/gyroData/magnetometerData, not the fused
        // CMDeviceMotion — U1 does its own AHRS fusion; feeding it Apple's
        // already-fused orientation would double-fuse (per U4's Approach).
        motionManager.startAccelerometerUpdates(to: .main) { [weak self] data, _ in
            guard let self, let data else { return }
            self.record(sensor: "accel", v: data.acceleration.tuple, uptime: data.timestamp)
        }
        motionManager.startGyroUpdates(to: .main) { [weak self] data, _ in
            guard let self, let data else { return }
            self.record(sensor: "gyro", v: data.rotationRate.tuple, uptime: data.timestamp)
        }
        motionManager.startMagnetometerUpdates(to: .main) { [weak self] data, _ in
            guard let self, let data else { return }
            self.record(sensor: "mag", v: data.magneticField.tuple, uptime: data.timestamp)
        }
    }

    @discardableResult
    func stop() -> URL? {
        motionManager.stopAccelerometerUpdates()
        motionManager.stopGyroUpdates()
        motionManager.stopMagnetometerUpdates()
        return writeTrialFile()
    }

    private func record(sensor: String, v: (Double, Double, Double), uptime: TimeInterval) {
        if startUptime == nil { startUptime = uptime }
        let tSeconds = uptime - (startUptime ?? uptime)
        let phoneTsMs = Int64((uptime * 1000).rounded())
        samples.append(Sample(tSeconds: tSeconds, sensor: sensor, v: v, phoneTsMs: phoneTsMs))
    }

    private func writeTrialFile() -> URL? {
        // Chronological order isn't guaranteed across sensor callback types
        // (per U6's Approach note) — sort before writing so downstream
        // replay_trial() gets a pre-sorted list, matching its documented contract.
        let sorted = samples.sorted { $0.tSeconds < $1.tSeconds }

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd-HHmmss"
        formatter.timeZone = TimeZone(identifier: "UTC")
        let stamp = formatter.string(from: Date())

        let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("trials", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let file = dir.appendingPathComponent("trial-\(stamp).json")

        var lines: [String] = ["["]
        for (index, s) in sorted.enumerated() {
            let v = "[\(s.v.0), \(s.v.1), \(s.v.2)]"
            let comma = index == sorted.count - 1 ? "" : ","
            lines.append(
                "  {\"t\": \(s.tSeconds), \"role\": \"proximal\", \"sensor\": \"\(s.sensor)\", " +
                "\"v\": \(v), \"phone_ts_ms\": \(s.phoneTsMs)}\(comma)"
            )
        }
        lines.append("]")

        do {
            try lines.joined(separator: "\n").write(to: file, atomically: true, encoding: .utf8)
            return file
        } catch {
            return nil
        }
    }
}

private extension CMAcceleration {
    var tuple: (Double, Double, Double) { (x, y, z) }
}
private extension CMRotationRate {
    var tuple: (Double, Double, Double) { (x, y, z) }
}
private extension CMMagneticField {
    var tuple: (Double, Double, Double) { (x, y, z) }
}

File: mobile-imu-core/harness-android/app/src/main/java/com/pendulastic/harness/SampleRecorder.kt
package com.pendulastic.harness

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import java.io.File
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone

/**
 * Throwaway capture-only recorder for U0 (see docs/plans/2026-08-21-001-feat-phone-imu-pendulum-app-plan.md).
 *
 * Writes one JSON array of samples per trial, each shaped as
 * {t, role, sensor, v, phone_ts_ms} to match imu_calibration_tuner.py's
 * replay_trial() input contract. `role` is fixed to "proximal" — that is
 * the role replay_trial()'s solo-fallback path (is_solo = not has_distal
 * and role == ROLE_PROXIMAL) expects for a single-phone capture; a "distal"
 * role with no proximal counterpart would NOT trigger that fallback.
 *
 * No calibration, no scoring, no participant model. This is not reused by
 * U3-U13 — see U0's plan entry for why.
 */
class SampleRecorder(private val context: Context) : SensorEventListener {

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val accel = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
    private val gyro = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
    private val mag = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)

    private val samples = mutableListOf<Sample>()
    private var recording = false
    private var startElapsedNanos = 0L

    val sampleCount: Int get() = samples.size

    private data class Sample(val tSeconds: Double, val sensor: String, val v: FloatArray, val phoneTsMs: Long)

    fun start() {
        samples.clear()
        // SensorEvent.timestamp is nanoseconds since boot (monotonic, not wall-clock) —
        // per KTD5/R1, this is the timestamp used for all dt math, never arrival time.
        startElapsedNanos = 0L
        recording = true
        // SENSOR_DELAY_FASTEST requests the highest rate the driver supports; U1/U2 need
        // ~100Hz, not the 800Hz CMBatchedSensorManager-equivalent path (KTD5) — FASTEST on
        // typical Android accel/gyro drivers lands in the low hundreds of Hz, which is fine.
        accel?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_FASTEST) }
        gyro?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_FASTEST) }
        mag?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_FASTEST) }
    }

    fun stop(): File {
        recording = false
        sensorManager.unregisterListener(this)
        return writeTrialFile()
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (!recording) return
        if (startElapsedNanos == 0L) startElapsedNanos = event.timestamp
        val tSeconds = (event.timestamp - startElapsedNanos) / 1_000_000_000.0
        val phoneTsMs = event.timestamp / 1_000_000L
        val sensorName = when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> "accel"
            Sensor.TYPE_GYROSCOPE -> "gyro"
            Sensor.TYPE_MAGNETIC_FIELD -> "mag"
            else -> return
        }
        samples.add(Sample(tSeconds, sensorName, event.values.copyOf(3), phoneTsMs))
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // Not scored by the harness — U1's stillness/bias logic is what cares about
        // signal quality, and this is deliberately not ported here.
    }

    private fun writeTrialFile(): File {
        // Chronological order isn't guaranteed across sensor callback types (per U6's
        // Approach note) — sort before writing so downstream replay_trial() gets a
        // pre-sorted list, matching its documented contract.
        val sorted = samples.sortedBy { it.tSeconds }
        val stamp = SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }.format(System.currentTimeMillis())
        // App-specific external storage — no runtime permission needed on any supported
        // API level, and still reachable via `adb pull` or a file manager for retrieval.
        val dir = File(context.getExternalFilesDir(null), "trials").apply { mkdirs() }
        val file = File(dir, "trial-$stamp.json")
        file.bufferedWriter().use { out ->
            out.write("[\n")
            sorted.forEachIndexed { index, s ->
                val v = "[${s.v[0]}, ${s.v[1]}, ${s.v[2]}]"
                out.write(
                    "  {\"t\": ${s.tSeconds}, \"role\": \"proximal\", \"sensor\": \"${s.sensor}\", " +
                        "\"v\": $v, \"phone_ts_ms\": ${s.phoneTsMs}}"
                )
                out.write(if (index == sorted.lastIndex) "\n" else ",\n")
            }
            out.write("]\n")
        }
        return file
    }
}
